<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="xxie6" />

<meta name="date" content="2024-02-29" />

<title>hnscc-analysis</title>

<script src="site_libs/header-attrs-2.26/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">EBCD_GBCD_comparison</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">hnscc-analysis</h1>
<h4 class="author">xxie6</h4>
<h4 class="date">2024-02-29</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2024-04-11
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>EBCD_GBCD_comparison/</code>
<span class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.1). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date
</a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git
repository, you know the exact version of the code that produced these
results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20240229code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(20240229)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20240229code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20240229)</code> was run prior to running
the code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrong924bd2d">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong> 924bd2d
</a>
</p>
</div>
<div id="strongRepositoryversionstrong924bd2d"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version 924bd2d.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .DS_Store
    Ignored:    .Rhistory

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were
made to the R Markdown (<code>analysis/hnscc-analysis.Rmd</code>) and
HTML (<code>docs/hnscc-analysis.html</code>) files. If you’ve configured
a remote Git repository (see <code>?wflow_git_remote</code>), click on
the hyperlinks in the table below to view the files as they were in that
past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
924bd2d
</td>
<td>
Annie Xie
</td>
<td>
2024-04-11
</td>
<td>
Add initial hnscc analysis to project
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<div id="introduction" class="section level1">
<h1>Introduction</h1>
<p>The Stephens Lab has proposed two different ways of performing
orthogonal matrix factorization. We aim to compare these two methods and
apply them to single cell RNA sequencing data.</p>
</div>
<div id="methods" class="section level1">
<h1>Methods</h1>
<div id="gbcd" class="section level2">
<h2>GBCD</h2>
<p>The first method is called Generalized Binary Covariance
Decomposition (GBCD). The GBCD model aims to produce the following
decomposition: <span class="math display">\[Y_{ij} \approx
\sum_{k=1}^{K} L_{ik}F_{jk}\]</span></p>
<p>where we make the following two assumptions:</p>
<ol style="list-style-type: decimal">
<li>Memberships <span class="math inline">\(L_{ik}\)</span> are
nonnegative and often nearly binary<br />
</li>
<li>The GEP signatures <span class="math inline">\(F_{\cdot k}\)</span>
are mutually orthogonal</li>
</ol>
<p>Note that these two assumptions imply that<br />
<span class="math display">\[YY^{T} \approx LF^{T}FL^{T} =
LDL^{T}\]</span></p>
<p>To fit the model, the GBCD method consists of the following
steps:</p>
<ol style="list-style-type: decimal">
<li>Finding <span class="math inline">\(L\)</span> and <span
class="math inline">\(D\)</span> in the second expression by performing
EBMF on <span class="math inline">\(YY^{T}\)</span><br />
</li>
<li>Estimating <span class="math inline">\(F\)</span> based off the
first expression by performing EBMF on <span
class="math inline">\(Y\)</span> with <span
class="math inline">\(L\)</span> fixed (at which point we do not force
the orthogonality assumption, so assumption two is a soft
assumption)</li>
</ol>
</div>
<div id="ebcd" class="section level2">
<h2>EBCD</h2>
<p>The second method is called Empirical Bayes Covariance Decomposition
(EBCD). For EBCD, the model is <span class="math display">\[X = ZL^{T}
+E\]</span> where <span class="math inline">\(Z\)</span> is a <span
class="math inline">\(N \times K\)</span> orthogonal matrix and <span
class="math inline">\(L\)</span> is <span class="math inline">\(P \times
K\)</span> matrix over which we put a sparsity-inducing prior. (In
regular PCA, we would also assume <span
class="math inline">\(L^{T}L\)</span> is diagonal, but for sparse PCA,
we replace that assumption with a sparsity penalization.) An equivalent
represent for EBCD is <span class="math display">\[X^{T}X \approx
LL^{T}.\]</span></p>
<p>Note that in GBCD, we have <span class="math inline">\(YY^{T} \approx
LL^{T}\)</span> and we assume <span class="math inline">\(F\)</span> is
orthogonal. Therefore, <span class="math inline">\(F\)</span> maps to
<span class="math inline">\(Z\)</span> in EBCD and <span
class="math inline">\(Y^{T}\)</span> maps to <span
class="math inline">\(X\)</span> in EBCD. So when applying EBCD, you may
need to take the transpose of the original data matrix.</p>
<p>To fit the model, the EBCD algorithm iterates the following sequence
of steps:<br />
1. Update <span class="math inline">\(g_k\)</span> and <span
class="math inline">\(q_k\)</span> using an EBNM solver<br />
2. Update <span class="math inline">\(Z\)</span> by setting <span
class="math inline">\(Z = Polar.U(X\bar{L})\)</span><br />
3. Update the precision, <span class="math inline">\(\tau\)</span></p>
</div>
</div>
<div id="application-to-hnscc-data" class="section level1">
<h1>Application to HNSCC Data</h1>
<p>In this section, we will apply both methods to single cell RNA
sequencing data. We will compare the factorizations recovered from both
methods, and we also will compare the run time. The dataset is a head
and neck squamous cell carcinoma (hnscc) dataset. The data was collected
by Puram et al. from primary tumors from 10 hnscc patients. The rows
correspond to samples and the columns correspond to genes. Our data
matrix is <span class="math inline">\(2176 \times 17113\)</span>.</p>
<p>First we load the needed R packages:</p>
<pre class="r"><code>library(ggplot2)
library(cowplot)
library(RColorBrewer)
library(ggrepel)
library(pheatmap)
library(gridExtra)
library(Matrix)
library(ebnm)
library(flashier)</code></pre>
<pre><code>Loading required package: magrittr</code></pre>
<pre class="r"><code>library(magrittr)
library(ashr)
library(irlba)
library(reshape2)
source(&quot;~/Documents/PhD 3/Research/EBCD/gbcd-workflow/code/fit_cov_ebnmf.R&quot;)</code></pre>
<p>We also define some functions here.<br />
This function produces a heatmap visualization of the GEP
memberships:</p>
<pre class="r"><code>loadings_visualization &lt;- function(info, fit_L, sample.names, subject_col, subtype_col, subjects_subset = NULL){
  if (is.null(subjects_subset) == FALSE){
    anno &lt;- data.frame(subject = info$subject[(info$subject %in% subjects_subset)],
                   subtype = info$subtype[(info$subject %in% subjects_subset)])
    anno_colors &lt;- list(subject = subject_col[subjects_subset],
                    subtype = subtype_col)
  }
  else{
    anno &lt;- data.frame(subject = info$subject,
                   subtype = info$subtype)
    anno_colors &lt;- list(subject = subject_col,
                    subtype = subtype_col)
  }
  rownames(anno) &lt;- sample.names
  cols &lt;- colorRampPalette(c(&quot;gray96&quot;,&quot;red&quot;))(50)
  brks &lt;- seq(min(fit_L), max(fit_L),length.out= 50)
  rows &lt;- order(anno$subject)
  p &lt;- pheatmap(fit_L[rows, -1], cluster_rows = FALSE, cluster_cols = FALSE,
          show_rownames = FALSE, annotation_row = anno, annotation_names_row = FALSE,
          annotation_colors = anno_colors, angle_col = 45, fontsize = 8, color = cols, breaks = brks, main = &quot;&quot;)
  print(p)
}</code></pre>
<p>This function produces a volcano plot visualization of the GEPs and
driving genes of each GEP:</p>
<pre class="r"><code>volcano_plot_gbcd &lt;- function(fit.gbcd, k){
pdat &lt;- data.frame(gene = rownames(fit.gbcd$F$lfc), 
                   lfc  = fit.gbcd$F$lfc[,k], 
                   z    = abs(fit.gbcd$F$z_score[,k]), 
                   lfsr = fit.gbcd$F$lfsr[,k],
                   stringsAsFactors = FALSE)
pdat &lt;- transform(pdat,lfsr = cut(lfsr, c(-1,0.001,0.01,0.05,Inf)))
rows  &lt;- with(pdat, which(!(abs(lfc) &gt; quantile(abs(lfc),0.999) | (z &gt; 10))))
pdat[rows, &quot;gene&quot;] &lt;- &quot;&quot;
p &lt;- ggplot(pdat, aes(x = lfc, y = z, color = lfsr, label = gene)) +
  geom_point() + 
  geom_text_repel(color = &quot;black&quot;, size = 2.3, segment.color = &quot;black&quot;,
                  segment.size = 0.25, min.segment.length = 0,
                  max.overlaps = Inf, na.rm = TRUE) +
  scale_color_manual(values = c(&quot;coral&quot;,&quot;orange&quot;,&quot;gold&quot;,&quot;deepskyblue&quot;)) +
  labs(x = &quot;log-fold change&quot;, y = &quot;|posterior z-score|&quot;) + 
  theme_cowplot(font_size = 10)
print(p)
}</code></pre>
<p>This is a function to classify GEPs as patient-specific
vs. shared:</p>
<pre class="r"><code>gep_classification &lt;- function(ebcd.L){
  num.gep &lt;- ncol(ebcd.L) - 1
  gep.classifications &lt;- rep(&#39;0&#39;, num.gep)
  for (i in 2:num.gep){
    table.k &lt;- as.matrix(table(ebcd.L[,i] &gt; 0.05, ebcd.L$subject))
    ratio.k &lt;- table.k[2,]/colSums(table.k) #proportion of cells in patient i that show expression in GEP k
    max.ratio &lt;- max(ratio.k[ratio.k!=max(ratio.k)])/max(ratio.k)
    if (max.ratio &gt; 0.2){
      gep.classifications[(i-1)] &lt;- &#39;shared&#39;
    }
    else{
      gep.classifications[(i-1)] &lt;- &#39;patient-specific&#39;
    }
  }
  return(gep.classifications)
}</code></pre>
<p>This function generates the tile plot visualization of a matrix:</p>
<pre class="r"><code># #Yusha&#39;s code for Figure 2C

GEP_tile_plot &lt;- function(F.pm, k.idx1){
### load in lfc estimates for shared GEPs identified by gbcd
  F.pm &lt;- fit.gbcd$F$lfc[, k.idx1+1]
  colnames(F.pm) &lt;- paste0(&quot;GEP&quot;, (k.idx1))

### make the effect size tile plots of the top 10 driving genes for each shared GEP
  n &lt;- 10
  geps &lt;- paste0(&quot;GEP&quot;, (k.idx1))
  genes &lt;- c()
  for (gep in geps) {
    f &lt;- F.pm[, gep]
    i &lt;- head(order(f, decreasing = TRUE), n = n)
    i &lt;- setdiff(i, genes)
    genes &lt;- c(genes, i)
  }
  dat &lt;- cbind(gene = rownames(F.pm)[genes], as.data.frame(F.pm[genes, geps]))
  dat &lt;- transform(dat, gene = factor(gene, gene))
  rownames(dat) &lt;- NULL
  dat &lt;- melt(dat)
  names(dat) &lt;- c(&quot;gene&quot;,&quot;gep&quot;,&quot;lfc&quot;)
  dat &lt;- transform(dat, gep = factor(gep, rev(geps)), lfc = lfc)
  dat$lfc[dat$lfc &gt; 10] &lt;- 10
  dat$lfc[dat$lfc &lt; -10] &lt;- -10
  dat &lt;- transform(dat, lfc = cut(lfc, breaks = c(-10, -5, -2, -1, -0.5, 0, 0.5, 1, 2, 5, 10)))
  colors &lt;- colorRampPalette(c(&quot;blue&quot;,&quot;white&quot;,&quot;red&quot;))(10)
  p_gbcd &lt;- ggplot(dat, aes(x = gene, y = gep, fill = lfc)) + geom_tile() + scale_fill_manual(values = colors) +
    theme_cowplot(font_size = 8) + theme(axis.text.x = element_text(angle = 90, hjust = 1)) + labs(x = &quot;&quot;, y = &quot;&quot;, fill = &quot;LFC&quot;)
  print(p_gbcd)
}</code></pre>
<p>This RData file contains the hnscc dataset:</p>
<pre class="r"><code>load(&quot;~/Documents/PhD 3/Research/EBCD/gbcd-workflow/hnscc/hnscc.RData&quot;)</code></pre>
</div>
<div id="gbcd-analysis" class="section level1">
<h1>GBCD Analysis</h1>
<p>Yusha’s GBCD results are stored in this RData file. We will load the
results:</p>
<pre class="r"><code>load(&quot;~/Documents/PhD 3/Research/EBCD/gbcd-workflow/hnscc/hnscc_gbcd.RData&quot;)</code></pre>
<p>This is a heatmap visualizing the loadings for the GEPs:</p>
<pre class="r"><code>loadings_visualization(info, fit.gbcd$L, rownames(fit.gbcd$L), subject_col, subtype_col)</code></pre>
<p><img src="figure/hnscc-analysis.Rmd/unnamed-chunk-8-1.png" width="672" style="display: block; margin: auto;" /></p>
<div id="observations" class="section level2">
<h2>Observations</h2>
<p>From the loadings matrix, it looks like GBCD discovered 3
subtype-specific GEPs (one of which is also patient-specific), 7
patient-specific GEPs (including the subtype-specific GEP), and 19
shared GEPs. Yusha told me that she discarded some of the GEPs that
either were correlated to size factor or did not appear to have a
biological meaningful interpretation.</p>
</div>
<div id="interpretation-of-geps" class="section level2">
<h2>Interpretation of GEPs</h2>
<p>From visual inspection, we see that some GEPs correspond to
patient-specific effects. In addition, there are three GEPs which
correspond to the three subtypes. Lastly, other GEPs corresponded to
effects shared across multiple patients. Here is a full list of GEP
interpretations:</p>
<ol style="list-style-type: decimal">
<li>Classical subtype<br />
</li>
<li>Basal subtype<br />
</li>
<li>Atypical subtype/patient MEEI26 (two different colors correspond to
MEEI26 because there are two different tumors from this patient)<br />
</li>
<li>patient MEEI17<br />
</li>
<li>a shared GEP (though this GEP is more dense in the classical and
atypical subtypes)<br />
</li>
<li>shared GEP<br />
</li>
<li>patient MEEI25<br />
</li>
<li>shared GEP<br />
</li>
<li>patient MEEI6<br />
</li>
<li>shared GEP<br />
</li>
<li>patient MEEI18<br />
</li>
<li>shared GEP<br />
</li>
<li>shared GEP<br />
</li>
<li>shared GEP<br />
</li>
<li>shared GEP<br />
</li>
<li>shared GEP<br />
</li>
<li>shared GEP<br />
</li>
<li>shared GEP<br />
</li>
<li>shared GEP<br />
</li>
<li>shared GEP<br />
</li>
<li>shared GEP<br />
</li>
<li>shared GEP<br />
</li>
<li>shared GEP<br />
</li>
<li>shared GEP<br />
</li>
<li>patient MEEI5<br />
</li>
<li>shared GEP<br />
</li>
<li>patient MEEI22<br />
</li>
<li>shared GEP</li>
</ol>
</div>
<div id="classification-of-geps-as-patient-specific-vs-shared"
class="section level2">
<h2>Classification of GEPs as patient specific vs shared</h2>
<p>We have a quantitative method to classify GEPs as patient-specific
vs. shared:<br />
For a given GEP k,<br />
1. Choose a small epsilon, e.g. 0.05 (after you normalize the range of
loading values to [0,1]). We say that GEP k is expressed in cell <span
class="math inline">\(i\)</span> if <span class="math inline">\(l_{ik}
&gt; 0.05\)</span>.<br />
2. Calculate the proportion of cells expressing this GEP k for each
patient <span class="math inline">\(n=1,...N\)</span> which we denote by
<span class="math inline">\(\rho_1,...\rho_N\)</span>.<br />
3. Across the <span class="math inline">\(N\)</span> patients, calculate
the ratio of the second largest <span
class="math inline">\(\rho_n\)</span> to the largest <span
class="math inline">\(\rho_n\)</span>. If this ratio is close to 0, it
is considered patient-specific.</p>
<p>Rationale: if a GEP is patient-specific, say, expressed only in
patient n, that the proportion of cells expressing this GEP in any other
patient should be much smaller than the proportion for patient n. </p>
<pre class="r"><code>gbcd.L &lt;- data.frame(fit.gbcd$L)
gbcd.L$subject &lt;- info$sample.id
gbcd.gep_classifications &lt;- gep_classification(gbcd.L)
gbcd.k.idx1 &lt;- which(gbcd.gep_classifications == &#39;shared&#39;)
gbcd.k.idx2 &lt;- which(gbcd.gep_classifications == &#39;patient-specific&#39;)
gbcd.k.idx &lt;- c(gbcd.k.idx2, gbcd.k.idx1)</code></pre>
<pre class="r"><code>loadings_visualization(info, fit.gbcd$L[,c(1, gbcd.k.idx+1)], rownames(fit.gbcd$L), subject_col, subtype_col)</code></pre>
<p><img src="figure/hnscc-analysis.Rmd/unnamed-chunk-10-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="visualization-of-factor-matrix" class="section level2">
<h2>Visualization of factor matrix</h2>
<p>This is a tile plot visualization of the top 10 driving genes for
each shared GEP. We are trying to visualize (a subset of) the factor
matrix.</p>
<pre class="r"><code>GEP_tile_plot(fit.gbcd$F$lfc, gbcd.k.idx1)</code></pre>
<pre><code>Using gene as id variables</code></pre>
<p><img src="figure/hnscc-analysis.Rmd/unnamed-chunk-11-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="assessing-the-fit" class="section level2">
<h2>Assessing the fit</h2>
<pre class="r"><code>gbcd.fitted.values &lt;- fit.gbcd$L %*% t(fit.gbcd$F$lfc)</code></pre>
<pre class="r"><code>samp.vals &lt;- sample(c(1:prod(dim(Y))), size = 50000)</code></pre>
<pre class="r"><code>ggplot(data = NULL, aes(x = c(as.matrix(Y)[samp.vals]), y = c(gbcd.fitted.values)[samp.vals])) + geom_point() + xlab(&#39;Observed Values&#39;) + ylab(&#39;Fitted Values&#39;) + geom_abline(slope = 1, intercept = 0, color = &#39;red&#39;)</code></pre>
<p><img src="figure/hnscc-analysis.Rmd/unnamed-chunk-14-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>sum((Y-gbcd.fitted.values)^2)</code></pre>
<pre><code>[1] 51296854</code></pre>
</div>
</div>
<div id="ebcd-analysis" class="section level1">
<h1>EBCD Analysis</h1>
<p>I have the results from EBCD saved in a RData file:</p>
<pre class="r"><code>load(&quot;~/Documents/PhD 3/Research/EBCD/ebcd_hnscc_full.RData&quot;)</code></pre>
<p>This is a heatmap visualizing the loadings for the GEPs. In GBCD, the
loadings are scaled so that the maximum membership value for each GEP is
1. The EBCD method does not do this, so I apply the same scaling
transformation to the loadings estimate from EBCD. This is a heatmap
visualizing the scaled loadings for the GEPs:</p>
<pre class="r"><code>ebcd.fit_full_L &lt;- ebcd.fit_full$EL
gep_colnames &lt;- c(&#39;Baseline&#39;, paste(&#39;GEP&#39;, c(1:(ncol(ebcd.fit_full_L) - 1))))</code></pre>
<pre class="r"><code>#scaled version
ebcd.fit_full_L_scaled &lt;- t(t(ebcd.fit_full_L)/apply(ebcd.fit_full_L,2, max))
colnames(ebcd.fit_full_L_scaled) &lt;- gep_colnames
rownames(ebcd.fit_full_L_scaled) &lt;- rownames(Y)

loadings_visualization(info, ebcd.fit_full_L_scaled, rownames(Y), subject_col, subtype_col)</code></pre>
<p><img src="figure/hnscc-analysis.Rmd/unnamed-chunk-18-1.png" width="672" style="display: block; margin: auto;" /></p>
<div id="observations-1" class="section level2">
<h2>Observations</h2>
<p>From the loadings matrix, it seems that EBCD uncovered two
subtype-specific GEPs (one of which is also a patient-specific GEP),
eight patient-specific GEPs, and 19 shared GEPs. Something interesting
is EBCD uncovered one more patient-specific GEP than GBCD. Another
interesting observation is EBCD uncovered one fewer subtype-specific GEP
than GBCD. In addition, some of the shared GEPs from EBCD are most dense
in one or two of the subtypes – I’m not sure if this means that coupled
subtypes are grouped together in some way or if there’s some
identifiability issues that led to this coupling.</p>
</div>
<div id="interpretation-of-geps-1" class="section level2">
<h2>Interpretation of GEPs</h2>
<p>Here is a full list of GEP interpretations:</p>
<ol style="list-style-type: decimal">
<li>patient MEEI20<br />
</li>
<li>Atypical subtype/patient MEEI26 (two different colors correspond to
MEEI26 because there are two different tumors from this patient)<br />
</li>
<li>patient MEEI6<br />
</li>
<li>shared GEP<br />
</li>
<li>patient MEEI17<br />
</li>
<li>shared GEP<br />
</li>
<li>shared GEP (but it is most dense in the classical and atypical
subtypes)<br />
</li>
<li>patient MEEI25<br />
</li>
<li>patient MEEI18<br />
</li>
<li>shared GEP<br />
</li>
<li>shared GEP (but it is most dense in the classical and atypical
subtypes)<br />
</li>
<li>patient MEEI28<br />
</li>
<li>primarily classical subtype<br />
</li>
<li>shared GEP (but most dense in the basal subtype)<br />
</li>
<li>shared GEP<br />
</li>
<li>shared GEP<br />
</li>
<li>shared GEP<br />
</li>
<li>shared GEP<br />
</li>
<li>patient MEEI5<br />
</li>
<li>shared GEP<br />
</li>
<li>shared GEP<br />
</li>
<li>shared GEP (but it is most dense in the classical and atypical
subtypes)<br />
</li>
<li>shared GEP<br />
</li>
<li>shared GEP<br />
</li>
<li>shared GEP<br />
</li>
<li>shared GEP<br />
</li>
<li>shared GEP<br />
</li>
<li>shared GEP</li>
</ol>
</div>
<div id="classification-of-geps-as-patient-specific-vs-shared-1"
class="section level2">
<h2>Classification of GEPs as patient specific vs shared</h2>
<pre class="r"><code># ### specify the indices of shared and patient-specific GEPs respectively
ebcd.L &lt;- data.frame(ebcd.fit_full_L_scaled)
ebcd.L$subject &lt;- info$sample.id
ebcd.gep_classifications &lt;- gep_classification(ebcd.L)
ebcd.k.idx1 &lt;- which(ebcd.gep_classifications == &#39;shared&#39;)
ebcd.k.idx2 &lt;- which(ebcd.gep_classifications == &#39;patient-specific&#39;)
ebcd.k.idx &lt;- c(ebcd.k.idx2, ebcd.k.idx1)</code></pre>
<p>Here is a visualization of the loadings matrix with the GEPs
separated by patient-specific and shared:</p>
<pre class="r"><code>loadings_visualization(info, ebcd.fit_full_L_scaled[,c(1, (ebcd.k.idx+1))], rownames(Y), subject_col, subtype_col)</code></pre>
<p><img src="figure/hnscc-analysis.Rmd/unnamed-chunk-20-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="visualization-of-factor-matrix-1" class="section level2">
<h2>Visualization of factor matrix</h2>
<p>The EBCD output returns the <span class="math inline">\(Z\)</span>
estimate for an equivalent problem. However, to get the <span
class="math inline">\(Z\)</span> estimate for the original problem, you
must perform a transformation. I have the transformed Z saved in the
file:</p>
<pre class="r"><code>Z &lt;- readRDS(&quot;~/Documents/PhD 3/Research/EBCD/EBCD_hnscc_Z.rds&quot;)</code></pre>
<p>This is a tile plot visualization of the top 10 driving genes for
each shared GEP. We are trying to visualize (a subset of) the factor
matrix.</p>
<pre class="r"><code>### load in lfc estimates for shared GEPs identified by gbcd
GEP_tile_plot(Z, ebcd.k.idx1)</code></pre>
<pre><code>Using gene as id variables</code></pre>
<p><img src="figure/hnscc-analysis.Rmd/unnamed-chunk-22-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="gene-set-enrichment-analysis-of-factors"
class="section level2">
<h2>Gene Set Enrichment Analysis of factors</h2>
<p>Lastly, we did gene set enrichment analysis on the GEPs from EBCD.
Using gene lists, we run gene set enrichment analysis using code
provided by Yusha. The results are saved in these rds files. We also
have a visualization of the top 5 genesets that are over-represented in
the driving genes of the GEP. These plots were created using Yusha’s
code; these are not the credible sets that SuSie outputs.</p>
<pre class="r"><code>library(stringr)</code></pre>
<pre class="r"><code>#adapted from Yusha&#39;s code

# idx &lt;- 1
# ### display the gene set enrichment analysis result for the GEP signature
# dat.k &lt;- readRDS(paste0(&quot;~/Documents/PhD 3/Research/EBCD/gbcd-workflow/hnscc/gsea/GEP&quot;, idx, &quot;.rds&quot;))

gsea_plot &lt;- function(idx, dat.k){
  dat.k &lt;- dat.k[1:pmin(nrow(dat.k), 5),]
  dat.k$cluster &lt;- paste0(&quot;GEP&quot;, idx)
  dat.k$nlog10p &lt;- pmin(dat.k$nlog10pFishersExact, 15)
  dat.k$description &lt;- factor(dat.k$description, levels=dat.k$description)
  p2 &lt;- ggplot(dat.k, aes_string(x = &quot;description&quot;, y = &quot;cluster&quot;, size = &quot;nlog10pFishersExact&quot;)) + scale_x_discrete(labels = function(x) str_wrap(x, width = 20))
  p2 &lt;- p2 + geom_point(colour = &quot;red&quot;) + xlab(NULL) + ylab(NULL) + scale_size_continuous(range=c(2, 6)) + 
  guides(size = guide_legend(order = 1, title = TeX(r&quot;($-\log_{10}(P)$)&quot;))) + ggtitle(paste0(&quot;Geneset Over-representation Analysis - GEP&quot;, idx)) +
  theme(axis.text.x = element_text(size = 15, angle = 90, vjust = 0.5, hjust=1), legend.title = element_text(size = 18), plot.title = element_text(hjust = 0.5, size = 18), axis.text.y = element_blank(), axis.ticks.y = element_blank())
  print(p2)
}</code></pre>
<pre class="r"><code>library(gtools)
library(latex2exp)</code></pre>
<pre class="r"><code>gsea.k.idx1 &lt;- c(4,6,7,10,11,14,15,16,17,18,20,21,22,23,24,25,26,27,28) #shared GEPs
files &lt;- Sys.glob(&#39;~/Documents/PhD 3/Research/EBCD/EBCD_gsea/factor*&#39;, dirmark = FALSE)
files &lt;- gtools::mixedsort(files)
for (k in 1:length(files)){
  dat.k &lt;- readRDS(files[k])
  idx &lt;- gsea.k.idx1[k]
  gsea_plot(idx, dat.k)
}</code></pre>
<pre><code>Warning: `aes_string()` was deprecated in ggplot2 3.0.0.
ℹ Please use tidy evaluation idioms with `aes()`.
ℹ See also `vignette(&quot;ggplot2-in-packages&quot;)` for more information.
This warning is displayed once every 8 hours.
Call `lifecycle::last_lifecycle_warnings()` to see where this warning was
generated.</code></pre>
<p><img src="figure/hnscc-analysis.Rmd/unnamed-chunk-26-1.png" width="672" style="display: block; margin: auto;" /><img src="figure/hnscc-analysis.Rmd/unnamed-chunk-26-2.png" width="672" style="display: block; margin: auto;" /><img src="figure/hnscc-analysis.Rmd/unnamed-chunk-26-3.png" width="672" style="display: block; margin: auto;" /><img src="figure/hnscc-analysis.Rmd/unnamed-chunk-26-4.png" width="672" style="display: block; margin: auto;" /><img src="figure/hnscc-analysis.Rmd/unnamed-chunk-26-5.png" width="672" style="display: block; margin: auto;" /><img src="figure/hnscc-analysis.Rmd/unnamed-chunk-26-6.png" width="672" style="display: block; margin: auto;" /><img src="figure/hnscc-analysis.Rmd/unnamed-chunk-26-7.png" width="672" style="display: block; margin: auto;" /><img src="figure/hnscc-analysis.Rmd/unnamed-chunk-26-8.png" width="672" style="display: block; margin: auto;" /><img src="figure/hnscc-analysis.Rmd/unnamed-chunk-26-9.png" width="672" style="display: block; margin: auto;" /><img src="figure/hnscc-analysis.Rmd/unnamed-chunk-26-10.png" width="672" style="display: block; margin: auto;" /><img src="figure/hnscc-analysis.Rmd/unnamed-chunk-26-11.png" width="672" style="display: block; margin: auto;" /><img src="figure/hnscc-analysis.Rmd/unnamed-chunk-26-12.png" width="672" style="display: block; margin: auto;" /><img src="figure/hnscc-analysis.Rmd/unnamed-chunk-26-13.png" width="672" style="display: block; margin: auto;" /><img src="figure/hnscc-analysis.Rmd/unnamed-chunk-26-14.png" width="672" style="display: block; margin: auto;" /><img src="figure/hnscc-analysis.Rmd/unnamed-chunk-26-15.png" width="672" style="display: block; margin: auto;" /><img src="figure/hnscc-analysis.Rmd/unnamed-chunk-26-16.png" width="672" style="display: block; margin: auto;" /><img src="figure/hnscc-analysis.Rmd/unnamed-chunk-26-17.png" width="672" style="display: block; margin: auto;" /><img src="figure/hnscc-analysis.Rmd/unnamed-chunk-26-18.png" width="672" style="display: block; margin: auto;" /><img src="figure/hnscc-analysis.Rmd/unnamed-chunk-26-19.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="assessing-the-fit-1" class="section level2">
<h2>Assessing the fit</h2>
<pre class="r"><code>ebcd.fitted.values &lt;- Z %*% t(ebcd.fit_full$EL)</code></pre>
<pre class="r"><code>ggplot(data = NULL, aes(x = c(as.matrix(t(Y))[samp.vals]), y = c(ebcd.fitted.values)[samp.vals])) + geom_point() + xlab(&#39;Observed Values&#39;) + ylab(&#39;Fitted Values&#39;) + geom_abline(slope = 1, intercept = 0, color = &#39;red&#39;)</code></pre>
<p><img src="figure/hnscc-analysis.Rmd/unnamed-chunk-28-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>sum((t(Y)-ebcd.fitted.values)^2)</code></pre>
<pre><code>[1] 69044763</code></pre>
</div>
</div>
<div id="comparison-of-the-outputs" class="section level1">
<h1>Comparison of the outputs</h1>
<div id="comparing-the-loadings-matrices" class="section level2">
<h2>Comparing the Loadings matrices</h2>
<pre class="r"><code>loadings_visualization(info, fit.gbcd$L[,c(1, gbcd.k.idx+1)], rownames(fit.gbcd$L), subject_col, subtype_col)</code></pre>
<p><img src="figure/hnscc-analysis.Rmd/unnamed-chunk-30-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>loadings_visualization(info, ebcd.fit_full_L_scaled[,c(1, (ebcd.k.idx+1))], rownames(Y), subject_col, subtype_col)</code></pre>
<p><img src="figure/hnscc-analysis.Rmd/unnamed-chunk-31-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="computing-correlation-of-the-loadings-estimates"
class="section level2">
<h2>Computing correlation of the loadings estimates</h2>
<p>We first analyze the concordance between the loadings matrices from
GBCD and EBCD. We compute the correlations between the columns of the
GBCD loadings estimate and the columns of the EBCD loadings
estimate.</p>
<pre class="r"><code>correlation_matrix &lt;- cor(fit.gbcd$L, ebcd.fit_full_L)</code></pre>
<p>For each column of the EBCD loadings estimate, we identify the
corresponding column of the GBCD loadings estimate that is most highly
correlated.</p>
<pre class="r"><code>max_correlation2 &lt;- apply(correlation_matrix, 2, FUN = max)
return_gep &lt;- function(x){
  gep_index &lt;- which.max(x)
  if (gep_index == 1){
    gep_name &lt;- &#39;Baseline&#39;
  }
  else{
    gep_name &lt;- paste(&#39;GEP&#39;,(gep_index - 1))
  }
  return(gep_name)
}
max_gbcd_gep &lt;- apply(correlation_matrix, 2, FUN = return_gep)</code></pre>
<pre class="r"><code>gep_colnames &lt;- c(&#39;Baseline&#39;, paste(&#39;GEP&#39;, c(1:(ncol(ebcd.fit_full_L) - 1))))
correlation_info1.2 &lt;- data.frame(max_gbcd_gep, max_correlation2, row.names = gep_colnames)
print(correlation_info1.2[order(correlation_info1.2$max_correlation2, decreasing = TRUE), ])</code></pre>
<pre><code>         max_gbcd_gep max_correlation2
Baseline     Baseline        0.9926223
GEP 3           GEP 9        0.9904580
GEP 8           GEP 7        0.9892136
GEP 5           GEP 4        0.9872016
GEP 2           GEP 3        0.9817010
GEP 1           GEP 1        0.9639916
GEP 19         GEP 25        0.9587368
GEP 9          GEP 11        0.8742653
GEP 21         GEP 14        0.8321674
GEP 11          GEP 1        0.8241774
GEP 28         GEP 12        0.7969180
GEP 4           GEP 8        0.7809100
GEP 13          GEP 1        0.7604189
GEP 7           GEP 1        0.6916835
GEP 22          GEP 5        0.6531408
GEP 18          GEP 6        0.6044451
GEP 17          GEP 2        0.5863873
GEP 25         GEP 26        0.5707728
GEP 14         GEP 28        0.5428904
GEP 6           GEP 6        0.5393635
GEP 10         GEP 13        0.5257710
GEP 26         GEP 18        0.5224878
GEP 23          GEP 2        0.5006992
GEP 24          GEP 2        0.4484873
GEP 20         GEP 16        0.4418414
GEP 12         GEP 23        0.4354294
GEP 15         GEP 27        0.3895859
GEP 16         GEP 16        0.3786894
GEP 27          GEP 5        0.3637642</code></pre>
</div>
<div id="computing-correlations-of-the-factor-estimates"
class="section level2">
<h2>Computing correlations of the factor estimates</h2>
<p>Next, we analyze the concordance between the factor matrices from
GBCD and EBCD. We follow the same steps.</p>
<pre class="r"><code>factor_correlation_matrix &lt;- cor(fit.gbcd$F$lfc, Z)</code></pre>
<p>For each column of the EBCD factors estimate, we identify the
corresponding column of the GBCD factors estimate that is most highly
correlated.</p>
<pre class="r"><code>factor_max_correlation2 &lt;- apply(factor_correlation_matrix, 2, FUN = max)
max_gbcd_factor_gep &lt;- apply(factor_correlation_matrix, 2, FUN = return_gep)</code></pre>
<pre class="r"><code># gep_colnames &lt;- c(&#39;Baseline&#39;, paste(&#39;GEP&#39;, c(1:(ncol(ebcd.fit_full_L) - 1))))
correlation_info2.2 &lt;- data.frame(max_gbcd_factor_gep, factor_max_correlation2, row.names = gep_colnames)
print(correlation_info2.2[order(correlation_info2.2$factor_max_correlation2, decreasing = TRUE), ])</code></pre>
<pre><code>         max_gbcd_factor_gep factor_max_correlation2
Baseline            Baseline               0.9738124
GEP 8                  GEP 7               0.9083285
GEP 2                  GEP 3               0.8586369
GEP 5                  GEP 4               0.8491722
GEP 19                GEP 25               0.8375469
GEP 3                  GEP 9               0.8147019
GEP 21                GEP 14               0.7829034
GEP 9                 GEP 11               0.7782279
GEP 1                  GEP 1               0.7303190
GEP 28                GEP 12               0.6570628
GEP 4                  GEP 8               0.6255442
GEP 18                 GEP 6               0.5357453
GEP 25                GEP 26               0.4956990
GEP 14                GEP 28               0.4780175
GEP 10                GEP 13               0.4615861
GEP 6                  GEP 6               0.4097887
GEP 15                GEP 27               0.4072208
GEP 26                GEP 18               0.4018246
GEP 24                GEP 20               0.3878383
GEP 17                GEP 28               0.3725357
GEP 20                GEP 24               0.3650022
GEP 16                GEP 16               0.3231424
GEP 22                 GEP 1               0.3118718
GEP 13                GEP 20               0.3110118
GEP 27                GEP 15               0.3008144
GEP 12                GEP 23               0.2936606
GEP 23                GEP 17               0.2911204
GEP 7                  GEP 5               0.2810376
GEP 11                GEP 21               0.2268515</code></pre>
</div>
<div id="looking-at-least-concordant-geps" class="section level2">
<h2>Looking at least concordant GEPs</h2>
<p>The least correlated GEP signatures (with respect to the loadings)
from the EBCD output are GEPs 10, 26, 23, 24, 20, 12, 15, 16, 27.</p>
<p>These are the top significantly enriched gene sets for those
GEPs:<br />
1. GEP10 - Neutrophil degranulation<br />
2. GEP26 - Genes up-regulated in response to alpha interferon
proteins<br />
3. GEP23 - Interferon alpha/beta signaling<br />
4. GEP24 - Eukaryotic Translation Elongation<br />
5. GEP20 - Genes regulated by NF-kB in response to TNF [GeneID =
7124]<br />
6. GEP17 - Genes defining epithelial-mesechymal transition, as in wound
healing, fibrosis and metastasis<br />
7. GEP12 - Patient specific – MEEI28 (didn’t do gene set analysis on
this)<br />
8. GEP15 - Formation of the cornified envelope<br />
9. GEP6 - Genes up-regulated in response to low oxygen levels
(hypoxia)<br />
10. GEP27 - Genes up-regulated in response to low oxygen levels
(hypoxia)</p>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.3.2 (2023-10-31)
Platform: aarch64-apple-darwin20 (64-bit)
Running under: macOS Ventura 13.6

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRblas.0.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.11.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: America/Chicago
tzcode source: internal

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] latex2exp_0.9.6    gtools_3.9.5       stringr_1.5.1      reshape2_1.4.4    
 [5] irlba_2.3.5.1      ashr_2.2-63        flashier_1.0.7     magrittr_2.0.3    
 [9] ebnm_1.1-2         Matrix_1.6-5       gridExtra_2.3      pheatmap_1.0.12   
[13] ggrepel_0.9.5      RColorBrewer_1.1-3 cowplot_1.1.3      ggplot2_3.5.0     
[17] workflowr_1.7.1   

loaded via a namespace (and not attached):
 [1] softImpute_1.4-1  gtable_0.3.4      xfun_0.43         bslib_0.7.0      
 [5] processx_3.8.4    lattice_0.22-6    callr_3.7.6       vctrs_0.6.5      
 [9] tools_4.3.2       ps_1.7.6          generics_0.1.3    parallel_4.3.2   
[13] tibble_3.2.1      fansi_1.0.6       highr_0.10        pkgconfig_2.0.3  
[17] SQUAREM_2021.1    lifecycle_1.0.4   truncnorm_1.0-9   farver_2.1.1     
[21] compiler_4.3.2    git2r_0.33.0      munsell_0.5.1     getPass_0.2-4    
[25] httpuv_1.6.15     htmltools_0.5.8   sass_0.4.9        yaml_2.3.8       
[29] crayon_1.5.2      tidyr_1.3.1       later_1.3.2       pillar_1.9.0     
[33] jquerylib_0.1.4   whisker_0.4.1     cachem_1.0.8      trust_0.1-8      
[37] tidyselect_1.2.1  digest_0.6.35     stringi_1.8.3     purrr_1.0.2      
[41] dplyr_1.1.4       labeling_0.4.3    splines_4.3.2     rprojroot_2.0.4  
[45] fastmap_1.1.1     grid_4.3.2        colorspace_2.1-0  cli_3.6.2        
[49] invgamma_1.1      utf8_1.2.4        withr_3.0.0       scales_1.3.0     
[53] promises_1.2.1    horseshoe_0.2.0   rmarkdown_2.26    httr_1.4.7       
[57] deconvolveR_1.2-1 evaluate_0.23     knitr_1.45        rlang_1.1.3      
[61] Rcpp_1.0.12       mixsqp_0.3-54     glue_1.7.0        rstudioapi_0.16.0
[65] jsonlite_1.8.8    plyr_1.8.9        R6_2.5.1          fs_1.6.3         </code></pre>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
