<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Annie Xie" />

<meta name="date" content="2024-04-22" />

<title>simulation-normal-data</title>

<script src="site_libs/header-attrs-2.26/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-6.4.2/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">EBCD_GBCD_comparison</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/xxie6/EBCD_GBCD_comparison">
    <span class="fab fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">simulation-normal-data</h1>
<h4 class="author">Annie Xie</h4>
<h4 class="date">2024-04-22</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2024-05-01
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>EBCD_GBCD_comparison/</code>
<span class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.1). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date
</a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git
repository, you know the exact version of the code that produced these
results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20240229code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(20240229)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20240229code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20240229)</code> was run prior to running
the code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomxxie6EBCDGBCDcomparisontreee192c9ebd15d50155fc59ec207bff888ed3e5fd1targetblanke192c9ea">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/xxie6/EBCD_GBCD_comparison/tree/e192c9ebd15d50155fc59ec207bff888ed3e5fd1" target="_blank">e192c9e</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomxxie6EBCDGBCDcomparisontreee192c9ebd15d50155fc59ec207bff888ed3e5fd1targetblanke192c9ea"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/xxie6/EBCD_GBCD_comparison/tree/e192c9ebd15d50155fc59ec207bff888ed3e5fd1" target="_blank">e192c9e</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .DS_Store
    Ignored:    .Rhistory
    Ignored:    data/.DS_Store

Untracked files:
    Untracked:  analysis/EBCD-subtype-exploration.Rmd
    Untracked:  code/EBCD subtype exploration full.Rmd
    Untracked:  data/iter1_normal_only_subtype_data.rds
    Untracked:  data/iter1_normal_subtype_only_ebcd.rds
    Untracked:  data/iter1_only_subtype_data.rds
    Untracked:  data/iter1_only_subtype_data2.rds
    Untracked:  data/iter1_only_subtype_data2_ebcd.rds
    Untracked:  data/iter1_subtype_only_ebcd.rds
    Untracked:  data/iter1_subtype_only_ebcd_true_init.RData

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were
made to the R Markdown
(<code>analysis/simulation-normal-data.Rmd</code>) and HTML
(<code>docs/simulation-normal-data.html</code>) files. If you’ve
configured a remote Git repository (see <code>?wflow_git_remote</code>),
click on the hyperlinks in the table below to view the files as they
were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/xxie6/EBCD_GBCD_comparison/blob/e192c9ebd15d50155fc59ec207bff888ed3e5fd1/analysis/simulation-normal-data.Rmd" target="_blank">e192c9e</a>
</td>
<td>
Annie Xie
</td>
<td>
2024-05-01
</td>
<td>
Made minor changes to simulation analyses
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/xxie6/EBCD_GBCD_comparison/553021af91016b3c2491f39af41db37ea4dc9ef3/docs/simulation-normal-data.html" target="_blank">553021a</a>
</td>
<td>
Annie Xie
</td>
<td>
2024-04-27
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/xxie6/EBCD_GBCD_comparison/blob/9150b8069133f70e1abb8c0d585e3173068bd7b3/analysis/simulation-normal-data.Rmd" target="_blank">9150b80</a>
</td>
<td>
Annie Xie
</td>
<td>
2024-04-27
</td>
<td>
Add normal data simulation
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<div id="introduction" class="section level1">
<h1>Introduction</h1>
<p>The Stephens Lab has proposed two different ways of performing
orthogonal matrix factorization, Generalized Binary Covariance
Decomposition (GBCD) and Empirical Bayes Covariance Decomposition
(EBCD). In this analysis, I want to compare these two methods on
simulated data with normal noise. The motivation for this analysis comes
from the observation that EBCD sometimes behaves unexpectedly on
simulated single-cell RNAseq data. I wondered if the unexpected behavior
was due to model misspecification and perhaps a lack of robustness for
this in the EBCD method. Therefore, I wanted to simulate data from the
true model that EBCD fits, and see if we still get any weird behavior. I
also want to compare how similar (and different) the EBCD and GBCD
results are.</p>
</div>
<div id="code-for-simulated-data" class="section level1">
<h1>Code for Simulated Data</h1>
<pre class="r"><code>##################################### simulate the single cell RNA-seq data for 20 replicates ##################################################
### simulate normal data
# note to self: I could take the L and F from previous data matrix and add normal random noise to LF&#39;
for(iter in 1:1){
  
  ### set the seed
  set.seed(iter)
  
  ### simulate L
  L &lt;- matrix(0, nrow=1600, ncol=11)
  L[1:800, 1] &lt;- 1
  L[801:1600, 2] &lt;- 1
  L[sample(1:nrow(L), 600, replace=FALSE), 3] &lt;- runif(600, min=0.4, max=2)
  L[1:200, 4] &lt;- 1
  L[201:400, 5] &lt;- 1
  L[401:600, 6] &lt;- 1
  L[601:800, 7] &lt;- 1
  L[801:1000, 8] &lt;- 1
  L[1001:1200, 9] &lt;- 1
  L[1201:1400, 10] &lt;- 1
  L[1401:1600, 11] &lt;- 1
  
  ### simulate F
  F &lt;- matrix(0, nrow=5000, ncol=11)
  F[1:75, 1] &lt;- pmax(rnorm(75, log2(3), 0.5), log2(1.5))
  F[76:150, 2] &lt;- pmax(rnorm(75, log2(3), 0.5), log2(1.5))
  F[251:500, 3] &lt;- pmax(rnorm(250, log2(3), 0.5), log2(1.5))
  F[501:1000, 4] &lt;- pmax(rnorm(500, log2(3), 0.5), log2(1.5))
  F[1001:1500, 5] &lt;- pmax(rnorm(500, log2(3), 0.5), log2(1.5))
  F[1501:2000, 6] &lt;- pmax(rnorm(500, log2(3), 0.5), log2(1.5))
  F[2001:2500, 7] &lt;- pmax(rnorm(500, log2(3), 0.5), log2(1.5))
  F[2501:3000, 8] &lt;- pmax(rnorm(500, log2(3), 0.5), log2(1.5))
  F[3001:3500, 9] &lt;- pmax(rnorm(500, log2(3), 0.5), log2(1.5))
  F[3501:4000, 10] &lt;- pmax(rnorm(500, log2(3), 0.5), log2(1.5))
  F[4001:4500, 11] &lt;- pmax(rnorm(500, log2(3), 0.5), log2(1.5))
  
  ### add normal noise
  E &lt;- matrix(rnorm(nrow(L)*nrow(F), mean = 0, sd = 1), ncol = nrow(F))
  Y = L %*% t(F) + E
  
  ### save the simulated data
  data &lt;- list(Y = Y, L = L, F = F)
  saveRDS(data, file=paste0(&quot;data/iter&quot;, iter, &quot;_normal_data.rds&quot;))
  rm(data, L, F, Y, E)
}</code></pre>
</div>
<div id="package-and-functions-for-analyses" class="section level1">
<h1>Package and Functions for Analyses</h1>
<pre class="r"><code>library(ggplot2)
library(cowplot)
library(RColorBrewer)
library(ggrepel)
library(pheatmap)
library(gridExtra)
#library(Seurat)
library(Matrix)
library(ebnm)
library(flashier)</code></pre>
<pre><code>Loading required package: magrittr</code></pre>
<pre class="r"><code>library(magrittr)
library(ashr)
library(irlba)
library(reshape2)

library(patchwork)</code></pre>
<pre><code>
Attaching package: &#39;patchwork&#39;</code></pre>
<pre><code>The following object is masked from &#39;package:cowplot&#39;:

    align_plots</code></pre>
<pre class="r"><code>library(fastTopics)
source(&quot;~/Documents/PhD 3/Research/EBCD/gbcd-workflow/code/fit_cov_ebnmf.R&quot;)</code></pre>
<pre class="r"><code>plot_heatmap &lt;- function(L, title = &quot;&quot;){
  ### define the color map
  cols &lt;- colorRampPalette(c(&quot;gray96&quot;, &quot;red&quot;))(49)
  brks &lt;- seq(min(L), max(L), length=50)
  
  plt &lt;- pheatmap(L, show_rownames = FALSE, show_colnames = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, color = cols, breaks = brks, main = title)
  return(plt)
}</code></pre>
<pre class="r"><code>source(&quot;~/Documents/PhD 3/Research/EBCD/ebcd_functions.R&quot;)
source(&quot;~/Documents/PhD 3/Research/EBCD/gbcd_functions.R&quot;)</code></pre>
</div>
<div id="simulated-dataset" class="section level1">
<h1>Simulated Dataset</h1>
<pre class="r"><code>### load in the simulated single cell data from this replicate
iter &lt;- 1
data &lt;- readRDS(paste0(&quot;data/iter&quot;, iter, &quot;_normal_data.rds&quot;))</code></pre>
<p>This is a heatmap of the loadings matrix:</p>
<pre class="r"><code>plot_heatmap(data$L)</code></pre>
<p><img src="figure/simulation-normal-data.Rmd/unnamed-chunk-6-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-6-1">
Past versions of unnamed-chunk-6-1.png
</button>
</p>
<div id="fig-unnamed-chunk-6-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/xxie6/EBCD_GBCD_comparison/blob/553021af91016b3c2491f39af41db37ea4dc9ef3/docs/figure/simulation-normal-data.Rmd/unnamed-chunk-6-1.png" target="_blank">553021a</a>
</td>
<td>
Annie Xie
</td>
<td>
2024-04-27
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>This is a heatmap of the factor matrix:</p>
<pre class="r"><code>plot_heatmap(data$F)</code></pre>
<p><img src="figure/simulation-normal-data.Rmd/unnamed-chunk-7-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-7-1">
Past versions of unnamed-chunk-7-1.png
</button>
</p>
<div id="fig-unnamed-chunk-7-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/xxie6/EBCD_GBCD_comparison/blob/553021af91016b3c2491f39af41db37ea4dc9ef3/docs/figure/simulation-normal-data.Rmd/unnamed-chunk-7-1.png" target="_blank">553021a</a>
</td>
<td>
Annie Xie
</td>
<td>
2024-04-27
</td>
</tr>
</tbody>
</table>
</div>
</div>
<pre class="r"><code>observed.vals &lt;- data$Y %*% t(data$Y)/ ncol(data$Y)</code></pre>
<p>This is a heatmap of the Gram matrix:</p>
<pre class="r"><code>plot_heatmap(observed.vals)</code></pre>
<p><img src="figure/simulation-normal-data.Rmd/unnamed-chunk-9-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="gbcd-analysis" class="section level1">
<h1>GBCD Analysis</h1>
<div id="hypothesis" class="section level2">
<h2>Hypothesis</h2>
<p>I think GBCD should be able to recover something close to the true
loadings estimate. However, it may not be exact since this loadings
matrix is not generated exactly from the model used to create the
loadings estimate. GBCD gets an estimate for the loadings matrix by
fitting <span class="math inline">\(LL^{T} + \sigma^2 I + E\)</span> to
the Gram matrix, where <span class="math inline">\(E\)</span> is
normally-distributed random noise. Regardless, I do think that GBCD
should recover one factor for each patient effect, one factor for each
subtype, and a factor corresponding to the shared GEP.</p>
</div>
<div id="analysis" class="section level2">
<h2>Analysis</h2>
<p>This is the code to run the GBCD analysis.</p>
<pre class="r"><code># I used Kmax = 16. I might want to try with different Kmax?
fit.gbcd &lt;- flash_fit_cov_ebnmf(Y = data$Y, Kmax = 16, prior = ebnm::ebnm_generalized_binary, thres = 0.9, extrapolate = FALSE)</code></pre>
<p>I have the results saved, so I will just load the results
directly.</p>
<pre class="r"><code>fit.gbcd &lt;- readRDS(paste0(&quot;data/iter&quot;, iter, &quot;_normal_data_gbcd.rds&quot;))</code></pre>
<p>This is a plot of estimate for <span
class="math inline">\(L\)</span>:</p>
<pre class="r"><code>plot_heatmap(fit.gbcd$L)</code></pre>
<p><img src="figure/simulation-normal-data.Rmd/unnamed-chunk-12-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>This is code to check if there are any factors that are numerically
zero. (This is something that I’ve seen happen in other estimates.) I
feel like it would make sense for GBCD to remove loadings that
correspond to numerically zero factors.</p>
<pre class="r"><code>sqrt(colSums(fit.gbcd$F$lfc^2))</code></pre>
<pre><code>Baseline     GEP1     GEP2     GEP3     GEP4     GEP5     GEP6     GEP7 
78.60203 64.56001 63.08175 63.10883 62.26083 62.06238 63.21376 62.56388 
    GEP8     GEP9    GEP10    GEP11    GEP12 
63.41924  0.00000  0.00000  0.00000  0.00000 </code></pre>
<p>Therefore are four factors that are numerically zero.</p>
<pre class="r"><code>non_zero_factor_idx &lt;- (sqrt(colSums(fit.gbcd$F$lfc^2)) &gt; 0)</code></pre>
<p>This is a heatmap of the loadings that correspond to non-zero
factors.</p>
<pre class="r"><code>plot_heatmap(fit.gbcd$L[,non_zero_factor_idx])</code></pre>
<p><img src="figure/simulation-normal-data.Rmd/unnamed-chunk-15-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="rescale-gbcd-loadings-estimate" class="section level2">
<h2>Rescale GBCD loadings estimate</h2>
<p>I’m loading in previously saved results.</p>
<pre class="r"><code>load(&quot;data/iter1_normal_data_gbcd_rescale.RData&quot;)</code></pre>
<p>This is the code to rescale the GBCD estimate. I’ve loaded in
previously saved results.</p>
<pre class="r"><code>fit.gbcd.rescale1 &lt;- flash_fit_cov_ebnmf_fit_laplace(Y = data$Y, Kmax = 16, prior = ebnm::ebnm_generalized_binary, thres = 0.9, extrapolate = FALSE, maxiter = 500, verbose = 1) 

fit.gbcd.rescale2 &lt;- flash_fit_cov_ebnmf_fit_L(dat = fit.gbcd.rescale1$dat, fit.gbcd.rescale1$fit.cov, Y=data$Y, Kmax=16, prior = ebnm::ebnm_generalized_binary, thres = 0.9, extrapolate = FALSE, maxiter = 500, verbose = 1)</code></pre>
<p><strong>LDF Method of Scaling:</strong></p>
<pre class="r"><code>fit.gbcd.rescale.ldf &lt;- ldf(fit.gbcd.rescale2$fit.cov, type = &#39;i&#39;)

fit.gbcd.rescale.L &lt;- fit.gbcd.rescale.ldf$L %*% diag(sqrt(fit.gbcd.rescale.ldf$D))

thres &lt;- 0.9
k.idx &lt;- which(fit.gbcd.rescale2$corr &gt; thres)
fit.gbcd.rescale.L &lt;- fit.gbcd.rescale.L[,fit.gbcd.rescale2$k.order][,k.idx]</code></pre>
</div>
<div id="assess-fit" class="section level2">
<h2>Assess Fit</h2>
<pre class="r"><code>plot_heatmap(fit.gbcd.rescale.L)</code></pre>
<p><img src="figure/simulation-normal-data.Rmd/unnamed-chunk-19-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>gbcd.rescaled.fitted.vals &lt;- fit.gbcd.rescale.L %*% t(fit.gbcd.rescale.L)</code></pre>
<pre class="r"><code>observed.vals &lt;- data$Y %*% t(data$Y)/ncol(data$Y)</code></pre>
<pre class="r"><code>sum((observed.vals - gbcd.rescaled.fitted.vals)^2) - sum((diag(observed.vals) - diag(gbcd.rescaled.fitted.vals))^2)</code></pre>
<pre><code>[1] 850.5104</code></pre>
<p>This is code to plot (a sub-sampple of) fitted values vs. observed
values:</p>
<pre class="r"><code>set.seed(3952)
diag_idx &lt;- seq(1, prod(dim(observed.vals)), length.out = ncol(observed.vals))
off_diag_idx &lt;- setdiff(c(1:prod(dim(observed.vals))), diag_idx) 
samp.vals &lt;- sample(off_diag_idx, size = 100000)</code></pre>
<pre class="r"><code>ggplot(data = NULL, aes(x = c(as.matrix(observed.vals))[samp.vals], y = c(gbcd.rescaled.fitted.vals)[samp.vals])) + geom_point() + xlab(&#39;Observed Values&#39;) + ylab(&#39;Fitted Values&#39;) + geom_abline(slope = 1, intercept = 0, color = &#39;red&#39;)</code></pre>
<p><img src="figure/simulation-normal-data.Rmd/unnamed-chunk-24-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-24-1">
Past versions of unnamed-chunk-24-1.png
</button>
</p>
<div id="fig-unnamed-chunk-24-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/xxie6/EBCD_GBCD_comparison/blob/553021af91016b3c2491f39af41db37ea4dc9ef3/docs/figure/simulation-normal-data.Rmd/unnamed-chunk-24-1.png" target="_blank">553021a</a>
</td>
<td>
Annie Xie
</td>
<td>
2024-04-27
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>This is a plot of the diagonal entries of the fitted values vs. the
diagonal entries of the observed values:</p>
<pre class="r"><code>ggplot(data = NULL, aes(x = diag(as.matrix(observed.vals)), y = diag(gbcd.rescaled.fitted.vals))) + geom_point() + xlab(&#39;Observed Values&#39;) + ylab(&#39;Fitted Values&#39;) + geom_abline(slope = 1, intercept = 0, color = &#39;red&#39;)</code></pre>
<p><img src="figure/simulation-normal-data.Rmd/unnamed-chunk-25-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-25-1">
Past versions of unnamed-chunk-25-1.png
</button>
</p>
<div id="fig-unnamed-chunk-25-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/xxie6/EBCD_GBCD_comparison/blob/553021af91016b3c2491f39af41db37ea4dc9ef3/docs/figure/simulation-normal-data.Rmd/unnamed-chunk-25-1.png" target="_blank">553021a</a>
</td>
<td>
Annie Xie
</td>
<td>
2024-04-27
</td>
</tr>
</tbody>
</table>
</div>
</div>
<pre class="r"><code>gbcd.rescaled.withdiag.fitted.vals &lt;- fit.gbcd.rescale.L %*% t(fit.gbcd.rescale.L) + diag(rep(fit.gbcd.rescale2$s2, nrow(fit.gbcd.rescale.L)))</code></pre>
<pre class="r"><code>sum((observed.vals - gbcd.rescaled.withdiag.fitted.vals)^2)</code></pre>
<pre><code>[1] 851.5753</code></pre>
<pre class="r"><code>samp.vals2 &lt;- sample(c(1:prod(dim(observed.vals))), size = 100000)</code></pre>
<pre class="r"><code>ggplot(data = NULL, aes(x = c(as.matrix(observed.vals))[samp.vals2], y = c(gbcd.rescaled.withdiag.fitted.vals)[samp.vals2])) + geom_point() + xlab(&#39;Observed Values&#39;) + ylab(&#39;Fitted Values&#39;) + geom_abline(slope = 1, intercept = 0, color = &#39;red&#39;)</code></pre>
<p><img src="figure/simulation-normal-data.Rmd/unnamed-chunk-29-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-29-1">
Past versions of unnamed-chunk-29-1.png
</button>
</p>
<div id="fig-unnamed-chunk-29-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/xxie6/EBCD_GBCD_comparison/blob/553021af91016b3c2491f39af41db37ea4dc9ef3/docs/figure/simulation-normal-data.Rmd/unnamed-chunk-29-1.png" target="_blank">553021a</a>
</td>
<td>
Annie Xie
</td>
<td>
2024-04-27
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="correlation-of-gbcd-estimate-to-true-loadings-matrix"
class="section level2">
<h2>Correlation of GBCD estimate to true loadings matrix</h2>
<p>We compute the correlation of the GBCD estimate of the loadings
matrix to the true loadings matrix.</p>
<pre class="r"><code>correlation_GBCD_true &lt;- cor(fit.gbcd$L, data$L)
colnames(correlation_GBCD_true) &lt;- c(&#39;Subtype 1 GEP&#39;, &#39;Subtype 2 GEP&#39;, &#39;Shared GEP&#39;, paste(&#39;Patient&#39;, c(1:8)))</code></pre>
<pre class="r"><code>correlation_GBCD_true</code></pre>
<pre><code>         Subtype 1 GEP Subtype 2 GEP   Shared GEP    Patient 1   Patient 2
Baseline    0.01207465   -0.01207465  0.998912275  0.008945767  0.03680248
GEP1        0.37982839   -0.37982839 -0.024464455 -0.141532996 -0.14192586
GEP2        0.37890836   -0.37890836  0.006038898  0.998971986 -0.14292360
GEP3       -0.37934481    0.37934481 -0.052289530 -0.143347994 -0.14345743
GEP4       -0.37895413    0.37895413  0.006946076 -0.142988741 -0.14352903
GEP5        0.38065681   -0.38065681  0.035385728 -0.140710693  0.99873948
GEP6        0.37618535   -0.37618535 -0.002038857 -0.143490348 -0.14349035
GEP7       -0.37902707    0.37902707  0.027132895 -0.143335718 -0.14333572
GEP8       -0.37736176    0.37736176 -0.005940165 -0.141961860 -0.14285183
GEP9        0.95581290   -0.95581290  0.034248445  0.261405392  0.34698584
GEP10      -0.96242096    0.96242096 -0.036034216 -0.363760931 -0.36376093
GEP11       0.60135094   -0.60135094  0.013331759  0.177601361  0.42153181
GEP12       0.02500524   -0.02500524 -0.034653709 -0.092615837  0.47719121
            Patient 3   Patient 4  Patient 5    Patient 6   Patient 7
Baseline -0.004137443 -0.02335565 -0.0546153 -0.003665835  0.03497515
GEP1     -0.141236892  0.99894229 -0.1437184 -0.143417676 -0.14316738
GEP2     -0.143461587 -0.13973120 -0.1429801 -0.143461587 -0.14295236
GEP3     -0.142215117 -0.14449490  0.9988874 -0.140694757 -0.14340277
GEP4     -0.143099177 -0.14330784 -0.1434159 -0.139278057 -0.14324929
GEP5     -0.142063835 -0.14046594 -0.1447590 -0.143761227 -0.14383274
GEP6      0.999209828 -0.14349035 -0.1418876 -0.142880953 -0.14165704
GEP7     -0.143027907 -0.14333572 -0.1433357 -0.142677494  0.99901148
GEP8     -0.142851831 -0.14285183 -0.1428518  0.999072849 -0.14285183
GEP9      0.517490332  0.31917171 -0.3620789 -0.359921704 -0.36207893
GEP10    -0.363760931 -0.36376093  0.4371542  0.341777375  0.39422457
GEP11    -0.247325575  0.55734957 -0.2282290 -0.238932558 -0.21829080
GEP12    -0.150296775 -0.19647423  0.2899538 -0.117581808 -0.14691308
           Patient 8
Baseline  0.00505084
GEP1     -0.14394309
GEP2     -0.14346159
GEP3     -0.14127441
GEP4      0.99886801
GEP5     -0.14314608
GEP6     -0.14231315
GEP7     -0.13996321
GEP8     -0.14285183
GEP9     -0.36097371
GEP10     0.28188761
GEP11    -0.22370480
GEP12    -0.06326329</code></pre>
</div>
<div id="interpretation-of-gbcd-geps" class="section level2">
<h2>Interpretation of GBCD GEPs</h2>
<p>Based off of the loadings, factors 2-9 correspond strongly to patient
effects. Factor 1 has the highest correlation with the shared GEP from
the true loadings matrix. Factor 9 has the highest correlation with the
subtype 1 GEP. Factor 10 has the highest correlation with the subtype 2
GEP. However, what is surprising is factors 9 and 10 in the factor
matrix are numerically zero. When you take out the factors that are
numerically zero, then the only factors left are those that correspond
to the patient effects and the shared GEP from the true loadings
matrix.</p>
</div>
</div>
<div id="ebcd-analysis-with-true-number-of-factors"
class="section level1">
<h1>EBCD Analysis (with true number of factors)</h1>
<div id="hypothesis-1" class="section level2">
<h2>Hypothesis</h2>
<p>EBCD should be able to recover the true loadings matrix since the
data was generated from the model that EBCD fits. Therefore, I
hypothesize that EBCD will find one factor for each patient effect, one
factor for each subtype, and a factor corresponding to the shared
GEP.</p>
</div>
<div id="analysis-1" class="section level2">
<h2>Analysis</h2>
<p>I’m loading in previously saved results.</p>
<pre class="r"><code>fit.ebcd &lt;- readRDS(paste0(&quot;data/iter&quot;, iter, &quot;_normal_data_ebcd.rds&quot;))</code></pre>
<p>This is the code to run the EBCD analysis. I’ve already loaded in the
saved results.</p>
<pre class="r"><code>set.seed(295)
fit.ebcd &lt;- ebcd(X = t(data$Y), Kmax = 12, ebnm_fn = ebnm::ebnm_generalized_binary)</code></pre>
<p>This is a plot of the scaled estimate of <span
class="math inline">\(L\)</span>. This estimate is scaled such that the
infinity norm for each column is 1, i.e. the maximum value for each
column is 1.</p>
<pre class="r"><code>plot_heatmap(t(t(fit.ebcd$EL)/apply(fit.ebcd$EL,2, max)))</code></pre>
<p><img src="figure/simulation-normal-data.Rmd/unnamed-chunk-34-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-34-1">
Past versions of unnamed-chunk-34-1.png
</button>
</p>
<div id="fig-unnamed-chunk-34-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/xxie6/EBCD_GBCD_comparison/blob/553021af91016b3c2491f39af41db37ea4dc9ef3/docs/figure/simulation-normal-data.Rmd/unnamed-chunk-34-1.png" target="_blank">553021a</a>
</td>
<td>
Annie Xie
</td>
<td>
2024-04-27
</td>
</tr>
</tbody>
</table>
</div>
</div>
<pre class="r"><code>ebcd.fitted.vals &lt;- fit.ebcd$EL %*% t(fit.ebcd$EL)</code></pre>
<pre class="r"><code>sum((observed.vals - ebcd.fitted.vals)^2)</code></pre>
<pre><code>[1] 3629.386</code></pre>
<pre class="r"><code>sum((observed.vals - ebcd.fitted.vals)^2) - sum((diag(observed.vals) - diag(ebcd.fitted.vals))^2)</code></pre>
<pre><code>[1] 2043.179</code></pre>
<p>This is a plot of (a subset of) the fitted values vs. observed
values:</p>
<pre class="r"><code>ggplot(data = NULL, aes(x = c(as.matrix(observed.vals))[samp.vals], y = c(ebcd.fitted.vals)[samp.vals])) + geom_point() + xlab(&#39;Observed Values&#39;) + ylab(&#39;Fitted Values&#39;) + geom_abline(slope = 1, intercept = 0, color = &#39;red&#39;)</code></pre>
<p><img src="figure/simulation-normal-data.Rmd/unnamed-chunk-38-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-38-1">
Past versions of unnamed-chunk-38-1.png
</button>
</p>
<div id="fig-unnamed-chunk-38-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/xxie6/EBCD_GBCD_comparison/blob/553021af91016b3c2491f39af41db37ea4dc9ef3/docs/figure/simulation-normal-data.Rmd/unnamed-chunk-38-1.png" target="_blank">553021a</a>
</td>
<td>
Annie Xie
</td>
<td>
2024-04-27
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>This is a plot of the diagonal entries of the fitted values vs. the
diagonal entries of the observed values:</p>
<pre class="r"><code>ggplot(data = NULL, aes(x = diag(as.matrix(observed.vals)), y = diag(ebcd.fitted.vals))) + geom_point() + xlab(&#39;Observed Values&#39;) + ylab(&#39;Fitted Values&#39;) + geom_abline(slope = 1, intercept = 0, color = &#39;red&#39;)</code></pre>
<p><img src="figure/simulation-normal-data.Rmd/unnamed-chunk-39-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-39-1">
Past versions of unnamed-chunk-39-1.png
</button>
</p>
<div id="fig-unnamed-chunk-39-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/xxie6/EBCD_GBCD_comparison/blob/553021af91016b3c2491f39af41db37ea4dc9ef3/docs/figure/simulation-normal-data.Rmd/unnamed-chunk-39-1.png" target="_blank">553021a</a>
</td>
<td>
Annie Xie
</td>
<td>
2024-04-27
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>This is a plot of the progression of the objective function</p>
<pre class="r"><code>ggplot(data = NULL, aes(x = c(1:length(fit.ebcd$vec.obj)), y = fit.ebcd$vec.obj)) + geom_line()</code></pre>
<p><img src="figure/simulation-normal-data.Rmd/unnamed-chunk-40-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-40-1">
Past versions of unnamed-chunk-40-1.png
</button>
</p>
<div id="fig-unnamed-chunk-40-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/xxie6/EBCD_GBCD_comparison/blob/553021af91016b3c2491f39af41db37ea4dc9ef3/docs/figure/simulation-normal-data.Rmd/unnamed-chunk-40-1.png" target="_blank">553021a</a>
</td>
<td>
Annie Xie
</td>
<td>
2024-04-27
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>This is the number of iterations that the backfit did before the
convergence criterion was satisfied:</p>
<pre class="r"><code>length(fit.ebcd$vec.obj)</code></pre>
<pre><code>[1] 168</code></pre>
</div>
<div id="correlation-of-ebcd-estimate-to-true-loadings-matrix"
class="section level2">
<h2>Correlation of EBCD estimate to true loadings matrix</h2>
<p>We compute the correlation of the EBCD estimate of the loadings
matrix to the true loadings matrix.</p>
<pre class="r"><code>correlation_EBCD_true &lt;- cor(fit.ebcd$EL, data$L)
colnames(correlation_EBCD_true) &lt;- c(&#39;Subtype 1 GEP&#39;, &#39;Subtype 2 GEP&#39;, &#39;Shared GEP&#39;, paste(&#39;Patient&#39;, c(1:8)))</code></pre>
<pre class="r"><code>correlation_EBCD_true</code></pre>
<pre><code>      Subtype 1 GEP Subtype 2 GEP   Shared GEP   Patient 1   Patient 2
 [1,]    0.01678365   -0.01678365  0.997783416  0.01296309  0.03922650
 [2,]    0.55972183   -0.55972183 -0.016858507 -0.21155496 -0.21155493
 [3,]    0.37780729   -0.37780729  0.006520447  0.99958414 -0.14279773
 [4,]   -0.37784138    0.37784138 -0.055355362 -0.14281062 -0.14281062
 [5,]   -0.37781572    0.37781572 -0.001507943 -0.14280092 -0.14280092
 [6,]   -0.37780084    0.37780084  0.007533974 -0.14279530 -0.14279530
 [7,]   -0.37782825    0.37782825  0.034283975 -0.14280565 -0.14280565
 [8,]    0.37757626   -0.37757626 -0.023219599 -0.14271041 -0.14271041
 [9,]    0.37780551   -0.37780551  0.034935134 -0.14279706  0.99957941
[10,]    0.37750112   -0.37750112 -0.001902382 -0.14268201 -0.14268201
[11,]   -0.05569523    0.05569523 -0.847204855  0.00565233 -0.04507800
[12,]   -0.05286861    0.05286861  0.824022779 -0.04140664  0.01397902
         Patient 3    Patient 4   Patient 5    Patient 6   Patient 7
 [1,]  0.001162852 -0.027977951 -0.05386227 -0.006340365  0.03121386
 [2,]  0.635821785  0.633507973 -0.21155497 -0.211554967 -0.21155497
 [3,] -0.142797735 -0.142797735 -0.14279773 -0.142797735 -0.14279773
 [4,] -0.142810617 -0.142810617  0.99967432 -0.142810617 -0.14281062
 [5,] -0.142800919 -0.142800919 -0.14280092  0.999606431 -0.14280092
 [6,] -0.142795297 -0.142795297 -0.14279530 -0.142795297 -0.14279530
 [7,] -0.142805654 -0.142805654 -0.14280565 -0.142805654  0.99963958
 [8,] -0.142710411  0.998972874 -0.14271041 -0.142710411 -0.14271041
 [9,] -0.142797059 -0.142797059 -0.14279706 -0.142797059 -0.14279706
[10,]  0.998774076 -0.142682011 -0.14268201 -0.142682011 -0.14268201
[11,] -0.034434522 -0.010343086  0.04545250  0.005267222  0.01245064
[12,] -0.061911043  0.009408842 -0.02316298  0.015810474  0.05866008
         Patient 8
 [1,]  0.003614292
 [2,] -0.211554967
 [3,] -0.142797735
 [4,] -0.142810617
 [5,] -0.142800919
 [6,]  0.999567079
 [7,] -0.142805654
 [8,] -0.142710411
 [9,] -0.142797059
[10,] -0.142682011
[11,]  0.021032920
[12,]  0.028622236</code></pre>
</div>
<div id="interpretation-of-ebcd-geps" class="section level2">
<h2>Interpretation of EBCD GEPs</h2>
<p>Based off of the loadings, factors 3-10 correspond strongly to
patient effects. Factor 1 has the highest correlation with the shared
GEP from the true loadings matrix with a correlation value of 0.998.</p>
</div>
</div>
<div id="comparing-gbcd-and-ebcd" class="section level1">
<h1>Comparing GBCD and EBCD</h1>
<div id="correlation-between-gbcd-and-ebcd-loadings-estimates"
class="section level2">
<h2>Correlation between GBCD and EBCD loadings estimates</h2>
<p>We analyze the concordance between the loadings matrices from GBCD
and EBCD. We compute the correlations between the columns of the GBCD
loadings estimate and the columns of the EBCD loadings estimate.</p>
<pre class="r"><code>correlation_L_estimates &lt;- cor(fit.gbcd$L, fit.ebcd$EL)</code></pre>
<p>For each column of the EBCD loadings estimate, we identify the
corresponding column of the GBCD loadings estimate that is most highly
correlated.</p>
<pre class="r"><code>max_corr &lt;- apply(correlation_L_estimates, 2, FUN = max)
return_gep &lt;- function(x){
  gep_index &lt;- which.max(x)
  if (gep_index == 1){
    gep_name &lt;- &#39;Baseline&#39;
  }
  else{
    gep_name &lt;- paste(&#39;GEP&#39;,(gep_index - 1))
  }
  return(gep_name)
}
max_gbcd_gep &lt;- apply(correlation_L_estimates, 2, FUN = return_gep)</code></pre>
<pre class="r"><code>gep_colnames &lt;- c(&#39;Baseline&#39;, paste(&#39;GEP&#39;, c(1:(ncol(fit.ebcd$EL) - 1))))
correlation_info &lt;- data.frame(max_gbcd_gep, max_corr, row.names = gep_colnames)
print(correlation_info[order(correlation_info$max_corr, decreasing = TRUE), ])</code></pre>
<pre><code>         max_gbcd_gep  max_corr
GEP 2           GEP 2 0.9993981
GEP 4           GEP 8 0.9993893
GEP 6           GEP 7 0.9992890
GEP 3           GEP 3 0.9992658
GEP 5           GEP 4 0.9992341
GEP 8           GEP 5 0.9991712
GEP 9           GEP 6 0.9989250
GEP 7           GEP 1 0.9987073
Baseline     Baseline 0.9976943
GEP 11       Baseline 0.8268212
GEP 1           GEP 1 0.6340588
GEP 10         GEP 10 0.0853049</code></pre>
</div>
<div id="observations" class="section level2">
<h2>Observations</h2>
<p>As previously mentioned, the GBCD method found factors that strongly
correspond with the shared GEP, patient-effect GEPs, and
subtype-specific GEPs of the true loadings matrix. However, the
subtype-specific GEPs are numerically zero in the factor matrix. It was
surprising to me that the GBCD method found factors that are numerically
zero since GBCD also assumes normally distributed noise – I’m not
exactly sure why it did that in this setting.</p>
<p>The EBCD method found factors that strongly correspond with the
shared GEP and patient-effect GEPs from the true loadings matrix.
However, the EBCD method did not find any factors that correspond to
subtype-specific effects; it found two additional shared effect factors
plus a factor corresponding to two patients. I wonder if it would find
subtype effects if we fit EBCD with more factors. I also wanted to note
that I have done different versions of this experiment, and EBCD did
find the subtype effects in those cases. Therefore, I guess EBCD can
find the subtype effects, but it may not necessarily always find them
(and the goal is to figure out when). In my other experiments, I think I
had a larger sample size and larger number of genes, so I wonder if the
larger sample size helps the method find the shared effects. The signal
to noise ratio in this setting may also make it more difficult to find
the subtype GEPs.</p>
<p>I also noticed that the EBCD estimate had a worse fit to the data
than the GBCD estimate. I think it might be because the GBCD estimate
added more factors while I limited the EBCD estimate to only 12 factors
(since the true number of factors is 11).</p>
</div>
</div>
<div id="experiment-use-gbcd-output-to-initialize-ebcd"
class="section level1">
<h1>Experiment: Use GBCD output to initialize EBCD</h1>
<div id="hypothesis-2" class="section level2">
<h2>Hypothesis</h2>
<p>The main difference between the EBCD estimate and GBCD estimate is
the presence of the subtype effects. I’m curious to see if the EBCD
backfit will retain the factors that correspond to subtype effects. I
think it should since the data is generated with subtype effects. Of
course, the subtype effects in the GBCD estimate correspond to factors
that are numerically zero. EBCD will not produce numerically zero
factors, so I’m guessing the backfit will adjust these factors such that
they are not numerically zero.</p>
</div>
<div id="analysis-2" class="section level2">
<h2>Analysis</h2>
<p>This is the code to run the analysis. I will load in previously saved
results.</p>
<pre class="r"><code>Z.init &lt;- PolarU(fit.ebcd$A%*%fit.gbcd.rescale.L)
fitted.Y &lt;- Z.init%*%t(fit.gbcd.rescale.L)
tau.est &lt;- prod(dim(fit.ebcd$A)) / sum((fit.ebcd$A - fitted.Y)^2)
ebcd_obj_init_rescaled &lt;- list(
    A = fit.ebcd$A, N = fit.ebcd$N, nrowA = fit.ebcd$nrowA,
    tau = tau.est, Z = Z.init, EL = fit.gbcd.rescale.L, ebnm_fn = ebnm::ebnm_generalized_binary
  )</code></pre>
<pre class="r"><code>fit.ebcd.gbcd.init_rescaled &lt;- ebcd_backfit(ebcd_obj_init_rescaled, maxiter = 2500)</code></pre>
<pre class="r"><code>load(&#39;data/iter1_normal_data_ebcd_gbcd_init.RData&#39;)</code></pre>
<pre class="r"><code>#plot initialization
plot_heatmap(fit.gbcd.rescale.L)</code></pre>
<p><img src="figure/simulation-normal-data.Rmd/unnamed-chunk-50-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-50-1">
Past versions of unnamed-chunk-50-1.png
</button>
</p>
<div id="fig-unnamed-chunk-50-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/xxie6/EBCD_GBCD_comparison/blob/553021af91016b3c2491f39af41db37ea4dc9ef3/docs/figure/simulation-normal-data.Rmd/unnamed-chunk-50-1.png" target="_blank">553021a</a>
</td>
<td>
Annie Xie
</td>
<td>
2024-04-27
</td>
</tr>
</tbody>
</table>
</div>
</div>
<pre class="r"><code>plot_heatmap(t(t(fit.gbcd.rescale.L)/apply(fit.gbcd.rescale.L,2, max)))</code></pre>
<p><img src="figure/simulation-normal-data.Rmd/unnamed-chunk-51-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-51-1">
Past versions of unnamed-chunk-51-1.png
</button>
</p>
<div id="fig-unnamed-chunk-51-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/xxie6/EBCD_GBCD_comparison/blob/553021af91016b3c2491f39af41db37ea4dc9ef3/docs/figure/simulation-normal-data.Rmd/unnamed-chunk-51-1.png" target="_blank">553021a</a>
</td>
<td>
Annie Xie
</td>
<td>
2024-04-27
</td>
</tr>
</tbody>
</table>
</div>
</div>
<pre class="r"><code>plot_heatmap(t(t(fit.ebcd.gbcd.init_rescaled$EL)/apply(fit.ebcd.gbcd.init_rescaled$EL,2, max)))</code></pre>
<p><img src="figure/simulation-normal-data.Rmd/unnamed-chunk-52-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-52-1">
Past versions of unnamed-chunk-52-1.png
</button>
</p>
<div id="fig-unnamed-chunk-52-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/xxie6/EBCD_GBCD_comparison/blob/553021af91016b3c2491f39af41db37ea4dc9ef3/docs/figure/simulation-normal-data.Rmd/unnamed-chunk-52-1.png" target="_blank">553021a</a>
</td>
<td>
Annie Xie
</td>
<td>
2024-04-27
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>This is code to check if there are any factors that are numerically
zero.</p>
<pre class="r"><code>sqrt(colSums(fit.ebcd.gbcd.init_rescaled$Z^2))</code></pre>
<pre><code> [1] 40 40 40 40 40 40 40 40 40 40 40 40 40</code></pre>
<p>This is the number of backfit iterations that were needed to reach
convergence</p>
<pre class="r"><code>print(length(fit.ebcd.gbcd.init_rescaled$vec.obj))</code></pre>
<pre><code>[1] 127</code></pre>
<pre class="r"><code>plot_heatmap(ebcd_backfit(ebcd_obj_init_rescaled, maxiter = 5)$EL)</code></pre>
<pre class="r"><code>ebcd.gbcd.rescaled.init.fitted.vals &lt;- fit.ebcd.gbcd.init_rescaled$EL %*% t(fit.ebcd.gbcd.init_rescaled$EL)</code></pre>
<pre class="r"><code>sum((observed.vals - ebcd.gbcd.rescaled.init.fitted.vals)^2)</code></pre>
<pre><code>[1] 2598.207</code></pre>
<pre class="r"><code>sum((observed.vals - ebcd.gbcd.rescaled.init.fitted.vals)^2) - sum((diag(observed.vals) - diag(ebcd.gbcd.rescaled.init.fitted.vals))^2)</code></pre>
<pre><code>[1] 1023.226</code></pre>
<p>This is a plot of (a subset of) the fitted values vs. observed
values:</p>
<pre class="r"><code>ggplot(data = NULL, aes(x = c(as.matrix(observed.vals))[samp.vals], y = c(ebcd.gbcd.rescaled.init.fitted.vals)[samp.vals])) + geom_point() + xlab(&#39;Observed Values&#39;) + ylab(&#39;Fitted Values&#39;) + geom_abline(slope = 1, intercept = 0, color = &#39;red&#39;)</code></pre>
<p><img src="figure/simulation-normal-data.Rmd/unnamed-chunk-59-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-59-1">
Past versions of unnamed-chunk-59-1.png
</button>
</p>
<div id="fig-unnamed-chunk-59-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/xxie6/EBCD_GBCD_comparison/blob/553021af91016b3c2491f39af41db37ea4dc9ef3/docs/figure/simulation-normal-data.Rmd/unnamed-chunk-59-1.png" target="_blank">553021a</a>
</td>
<td>
Annie Xie
</td>
<td>
2024-04-27
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>This is a plot of the diagonal entries of the fitted values vs. the
diagonal entries of the observed values:</p>
<pre class="r"><code>ggplot(data = NULL, aes(x = diag(as.matrix(observed.vals)), y = diag(ebcd.gbcd.rescaled.init.fitted.vals))) + geom_point() + xlab(&#39;Observed Values&#39;) + ylab(&#39;Fitted Values&#39;) + geom_abline(slope = 1, intercept = 0, color = &#39;red&#39;)</code></pre>
<p><img src="figure/simulation-normal-data.Rmd/unnamed-chunk-60-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-60-1">
Past versions of unnamed-chunk-60-1.png
</button>
</p>
<div id="fig-unnamed-chunk-60-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/xxie6/EBCD_GBCD_comparison/blob/553021af91016b3c2491f39af41db37ea4dc9ef3/docs/figure/simulation-normal-data.Rmd/unnamed-chunk-60-1.png" target="_blank">553021a</a>
</td>
<td>
Annie Xie
</td>
<td>
2024-04-27
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>This is a plot of the progression of the objective function</p>
<pre class="r"><code>ggplot(data = NULL, aes(x = c(2:length(fit.ebcd.gbcd.init_rescaled$vec.obj)), y = fit.ebcd.gbcd.init_rescaled$vec.obj[-1])) + geom_line()</code></pre>
<p><img src="figure/simulation-normal-data.Rmd/unnamed-chunk-61-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-61-1">
Past versions of unnamed-chunk-61-1.png
</button>
</p>
<div id="fig-unnamed-chunk-61-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/xxie6/EBCD_GBCD_comparison/blob/553021af91016b3c2491f39af41db37ea4dc9ef3/docs/figure/simulation-normal-data.Rmd/unnamed-chunk-61-1.png" target="_blank">553021a</a>
</td>
<td>
Annie Xie
</td>
<td>
2024-04-27
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="correlation-to-true-loadings-matrix" class="section level2">
<h2>Correlation to true loadings matrix</h2>
<pre class="r"><code>correlation_EBCD_GBCD_init_true &lt;- cor(fit.ebcd.gbcd.init_rescaled$EL, data$L)

colnames(correlation_EBCD_GBCD_init_true) &lt;- c(&#39;Subtype 1 GEP&#39;, &#39;Subtype 2 GEP&#39;, &#39;Shared GEP&#39;, paste(&#39;Patient&#39;, c(1:8)))</code></pre>
<pre class="r"><code>correlation_EBCD_GBCD_init_true</code></pre>
<pre><code>      Subtype 1 GEP Subtype 2 GEP   Shared GEP     Patient 1   Patient 2
 [1,]    0.01335991   -0.01335991  0.998848406  0.0095392943  0.03600186
 [2,]    0.37773345   -0.37773345 -0.025311493 -0.1427698239 -0.14276982
 [3,]    0.37770519   -0.37770519  0.006879548  0.9993140044 -0.14275914
 [4,]   -0.37780871    0.37780871 -0.056064943 -0.1427982693 -0.14279827
 [5,]   -0.37778319    0.37778319  0.006495756 -0.1427886259 -0.14278863
 [6,]    0.37771928   -0.37771928  0.035088950 -0.1427644674  0.99935127
 [7,]    0.37777642   -0.37777642 -0.001669574 -0.1427860644 -0.14278606
 [8,]   -0.37773538    0.37773538  0.031272798 -0.1427705557 -0.14277056
 [9,]   -0.37769711    0.37769711 -0.004741361 -0.1427560893 -0.14275609
[10,]    0.95525334   -0.95525334  0.075333025  0.3580530099  0.35934547
[11,]   -0.91700969    0.91700969  0.059374243 -0.3376133448 -0.35075970
[12,]    0.07886956   -0.07886956  0.149276597 -0.0005485288  0.07015467
[13,]    0.02801010   -0.02801010  0.116369460 -0.0043253198  0.09311573
          Patient 3   Patient 4   Patient 5   Patient 6   Patient 7
 [1,] -0.0002827247 -0.02506014 -0.05328813 -0.00369190  0.03267341
 [2,] -0.1427698240  0.99938877 -0.14276982 -0.14276982 -0.14276982
 [3,] -0.1427591435 -0.14275914 -0.14275914 -0.14275914 -0.14275914
 [4,] -0.1427982693 -0.14279827  0.99958789 -0.14279827 -0.14279827
 [5,] -0.1427886259 -0.14278863 -0.14278863 -0.14278863 -0.14278863
 [6,] -0.1427644674 -0.14276447 -0.14276447 -0.14276447 -0.14276447
 [7,]  0.9995024511 -0.14278606 -0.14278606 -0.14278606 -0.14278606
 [8,] -0.1427705557 -0.14277056 -0.14277056 -0.14277056  0.99939389
 [9,] -0.1427560893 -0.14275609 -0.14275609  0.99929262 -0.14275609
[10,]  0.3712078271  0.35560099 -0.38210051 -0.35268269 -0.32448085
[11,] -0.3475659077 -0.35044938  0.40770723  0.27092284  0.30249673
[12,] -0.0355054019  0.08513883 -0.02430053 -0.06845432 -0.04781965
[13,]  0.0072129500 -0.05365607  0.06359219 -0.07078268 -0.05505076
         Patient 8
 [1,]  0.004108336
 [2,] -0.142769824
 [3,] -0.142759143
 [4,] -0.142798269
 [5,]  0.999520381
 [6,] -0.142764467
 [7,] -0.142786064
 [8,] -0.142770556
 [9,] -0.142756089
[10,] -0.384943248
[11,]  0.405261532
[12,]  0.021334932
[13,]  0.019893957</code></pre>
</div>
<div id="observations-1" class="section level2">
<h2>Observations</h2>
<p>The EBCD backfit retained a lot of the GBCD estimate. In particular,
it kept factor 1 which highly corresponds to the the shared GEP from the
true loadings matrix. In addition, the estimate has all of the patient
effects as separate factors. The backfit kept most of the integrity of
the subtype-specific factors, but it did add some positive loadings to
samples not in the respective subtype. The backfit also made the
additional shared GEPs a lot more dense (with respect to the
loadings).</p>
<p>The fit of this estimate (disregarding the diagonal entries) is
better than the fit of the EBCD with greedy initialization estimate.
However, it is not as good as the GBCD estimate.</p>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.3.2 (2023-10-31)
Platform: aarch64-apple-darwin20 (64-bit)
Running under: macOS Sonoma 14.4.1

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRblas.0.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.11.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: America/Chicago
tzcode source: internal

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] fastTopics_0.6-142 patchwork_1.2.0    reshape2_1.4.4     irlba_2.3.5.1     
 [5] ashr_2.2-63        flashier_1.0.7     magrittr_2.0.3     ebnm_1.1-2        
 [9] Matrix_1.6-5       gridExtra_2.3      pheatmap_1.0.12    ggrepel_0.9.5     
[13] RColorBrewer_1.1-3 cowplot_1.1.3      ggplot2_3.5.1      workflowr_1.7.1   

loaded via a namespace (and not attached):
 [1] pbapply_1.7-2      rlang_1.1.3        git2r_0.33.0       horseshoe_0.2.0   
 [5] compiler_4.3.2     getPass_0.2-4      callr_3.7.6        vctrs_0.6.5       
 [9] quantreg_5.97      quadprog_1.5-8     stringr_1.5.1      pkgconfig_2.0.3   
[13] crayon_1.5.2       fastmap_1.1.1      mcmc_0.9-8         labeling_0.4.3    
[17] utf8_1.2.4         promises_1.3.0     rmarkdown_2.26     ps_1.7.6          
[21] MatrixModels_0.5-3 purrr_1.0.2        xfun_0.43          cachem_1.0.8      
[25] trust_0.1-8        jsonlite_1.8.8     progress_1.2.3     highr_0.10        
[29] later_1.3.2        parallel_4.3.2     prettyunits_1.2.0  R6_2.5.1          
[33] bslib_0.7.0        stringi_1.8.3      SQUAREM_2021.1     jquerylib_0.1.4   
[37] Rcpp_1.0.12        knitr_1.45         httpuv_1.6.15      splines_4.3.2     
[41] tidyselect_1.2.1   rstudioapi_0.16.0  yaml_2.3.8         processx_3.8.4    
[45] lattice_0.22-6     tibble_3.2.1       plyr_1.8.9         withr_3.0.0       
[49] coda_0.19-4.1      evaluate_0.23      Rtsne_0.17         survival_3.6-4    
[53] RcppParallel_5.1.7 pillar_1.9.0       whisker_0.4.1      plotly_4.10.4     
[57] softImpute_1.4-1   generics_0.1.3     rprojroot_2.0.4    invgamma_1.1      
[61] truncnorm_1.0-9    hms_1.1.3          munsell_0.5.1      scales_1.3.0      
[65] glue_1.7.0         lazyeval_0.2.2     tools_4.3.2        data.table_1.15.4 
[69] SparseM_1.81       fs_1.6.4           grid_4.3.2         tidyr_1.3.1       
[73] MCMCpack_1.7-0     colorspace_2.1-0   deconvolveR_1.2-1  cli_3.6.2         
[77] fansi_1.0.6        mixsqp_0.3-54      viridisLite_0.4.2  dplyr_1.1.4       
[81] uwot_0.1.16        gtable_0.3.5       sass_0.4.9         digest_0.6.35     
[85] farver_2.1.1       htmlwidgets_1.6.4  htmltools_0.5.8.1  lifecycle_1.0.4   
[89] httr_1.4.7         MASS_7.3-60.0.1   </code></pre>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
